velox/common/file/FileSystems.cpp

```diff
@@ -0,0 +1,106 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/file/FileSystems.h"
+#include "velox/common/base/Exceptions.h"
+#include "velox/common/file/File.h"
+#include "velox/core/Context.h"
+
+namespace facebook::velox::filesystems {
+
+constexpr std::string_view kFileScheme("file:");
+
+using RegisteredFileSystems = std::vector<std::pair<
+    std::function<bool(std::string_view)>,
+    std::function<std::shared_ptr<FileSystem>(std::shared_ptr<const Config>)>>>;
+
+RegisteredFileSystems& registeredFileSystems() {
+  // Meyers singleton.
+  static RegisteredFileSystems* fss = new RegisteredFileSystems();
+  return *fss;
+}
+
+void registerFileSystem(
+    std::function<bool(std::string_view)> schemeMatcher,
+    std::function<std::shared_ptr<FileSystem>(std::shared_ptr<const Config>)>
+        fileSystemGenerator) {
+  registeredFileSystems().emplace_back(schemeMatcher, fileSystemGenerator);
+}
+
+std::shared_ptr<FileSystem> getFileSystem(
+    std::string_view filename,
+    std::shared_ptr<const Config> properties) {
+  const auto& filesystems = registeredFileSystems();
+  for (const auto& p : filesystems) {
+    if (p.first(filename)) {
+      return p.second(properties);
+    }
+  }
+  VELOX_FAIL("No registered file system matched with filename '{}'", filename);
+}
+
+// Implement Local FileSystem.
+class LocalFileSystem : public FileSystem {
+ public:
+  LocalFileSystem(std::shared_ptr<const Config> config) : FileSystem(config) {}
+  virtual ~LocalFileSystem() {}
+  virtual std::string name() const override {
+    return "Local FS";
+  }
+  virtual std::unique_ptr<ReadFile> openFileForRead(
+      std::string_view path) override {
+    if (path.find(kFileScheme) == 0) {
+      return std::make_unique<LocalReadFile>(path.substr(kFileScheme.length()));
+    }
+    return std::make_unique<LocalReadFile>(path);
+  }
+  virtual std::unique_ptr<WriteFile> openFileForWrite(
+      std::string_view path) override {
+    if (path.find(kFileScheme) == 0) {
+      return std::make_unique<LocalWriteFile>(
+          path.substr(kFileScheme.length()));
+    }
+    return std::make_unique<LocalWriteFile>(path);
+  }
+
+  static std::function<bool(std::string_view)> schemeMatcher() {
+    // Note: presto behavior is to prefix local paths with 'file:'.
+    // Check for that prefix and prune to absolute regular paths as needed.
+    return [](std::string_view filename) {
+      return filename.find("/") == 0 || filename.find(kFileScheme) == 0;
+    };
+  }
+
+  static std::function<
+      std::shared_ptr<FileSystem>(std::shared_ptr<const Config>)>
+  fileSystemGenerator() {
+    return [](std::shared_ptr<const Config> properties) {
+      return std::make_shared<LocalFileSystem>(properties);
+    };
+  }
+};
+
+void registerLocalFileSystem() {
```

Comment:
@mbasmanova  I am now thinking if we should implicitly always register the LocalFileSystem if it is not explicitly registered? Isn't the Local FileSystem a dependency for Velox?
We could achieve this as follows. What do you think?
```
std::once_flag localFSRegistrationFlag;
void registerLocalFileSystem() {
  std::call_once(localFSRegistrationFlag, []() {
    registerFileSystem(LocalFileSystem::schemeMatcher(), LocalFileSystem::fileSystemGenerator());
  });
}
```

Reply:
My preference would be for the application using Velox to explicitly register a subset of file systems that are required.

Reply:
Fair enough. We can always make changes later if needed. This PR is ready for merge.

Reply:
@majetideepak Just to confirm, we are keeping registerAllFileSystems for now?

Reply:
Yes, my preference is to keep that for now.